GH-138: Fail fast when configured with invalid time partitioning properties

C0urante opened the pull request:
Addresses #138.

Depends on https://github.com/confluentinc/kafka-connect-bigquery/pull/163

Functional changes:
- A preflight check is made to ensure that the connector can't be configured to both use decorator syntax and create time-partitioned tables with a partitioning type other than `DAY`
- A runtime check is added to fail faster if the connector is configured to use decorator syntax and it detects that it will write to a table with a partitioning type other than `DAY`
- The logic for writing to non-`DAY`-partitioned tables with decorator syntax is removed as it serves no purpose

Testing changes:
- Some small unit test changes are added to verify the above changes
- A new integration test is added to cover a wide variety of configuration permutations related to time partitioning

I was initially optimistic that we could use the `_PARTITIONTIME` pseudocolumn to redirect rows to different partitions instead of using decorator syntax, but unfortunately, although this works for `INSERT` statements made in the BigQuery UI, it is not supported for streaming inserts made by the connector.

ncliang reviewed (approved):
Thanks @C0urante . LGTM with a couple questions and comments.

C0urante commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java:
```diff
@@ -757,6 +756,31 @@ private void checkBigQuerySchemaUpdateConfigs() {
+    )).orElse(true);
+    boolean createTables = getBoolean(TABLE_CREATE_CONFIG);
+
+    if (!decoratorSyntax || !createTables) {
```
> My understanding was that we will attempt to use decorator syntax to write to the table in `BigQuerySinkTaks::getRecordTable` even if create table was not specified.

This is correct; however, the time value specified for `timePartitioningType` only comes into play when automatically creating tables. If a table already exists, the connector will check for its partitioning type and adjust accordingly (in the case of this PR, by failing if decorator syntax is enabled and the time partitioning type isn't `DAY`).

I chose to fail only when table creation was enabled in order to avoid breaking existing connectors that might be mistakenly configured with a non-`DAY` partitioning type but without automatic table creation enabled. In that case the `timePartitioningType` property has no effect on the connector, so breaking because of how it's set seems unfair.

What are your thoughts?

C0urante commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/utils/PartitionedTableId.java:
```diff
@@ -203,19 +203,6 @@ public Builder setPartition(String partition) {
       return this;
     }
 
-    public Builder setHourPartition(long utcTime) {
```
Good call, will add.

C0urante commented:
@ncliang thanks for the review. I think your question about the preflight check for time partitioning type is a valid one; would you mind revisiting that when you have a chance? Would like to make sure we're on the same page before merging.

ncliang reviewed (approved):
LGTM. Thanks @C0urante

C0urante commented:
I've fixed the merge conflicts and updated the integration tests to pass now. In the process, it became apparent that some BigQuery behavior around writing to recently deleted-then-recreated tables has changed, so I've opened https://github.com/confluentinc/kafka-connect-bigquery/pull/163 and included its changes in this PR. Will wait for https://github.com/confluentinc/kafka-connect-bigquery/pull/163 to be approved before merging this PR.
